### PR TITLE
NuDB tidying and bugfix

### PR DESCRIPTION
**Pull request by vinniefalco**
Reopened

### CODE REVIEW

File: src/beast/beast/nudb/win32_file.h [hunk 1]
```diff
@@ -42,6 +42,9 @@
 # ifndef STRICT
 #  define STRICT
 # endif
+# ifndef WIN32_LEAN_AND_MEAN
```
Comment (rec) [inline src/beast/beast/nudb/win32_file.h]: This is already defined in _ten_ different places in the code - do we really need an eleventh?

File: src/beast/beast/nudb/store.h [hunk 2]
```diff
@@ -861,14 +862,11 @@ store<Hasher, Codec, File>::commit()
...
-    {
-        std::lock_guard<gentex> g (g_);
-        // Write new buckets to key file
-        for (auto const e : s_->c1)
-            e.second.write (s_->kf,
-                (e.first + 1) * s_->kh.block_size);
-    }
+    g_.finish();
+    // Write new buckets to key file
+    for (auto const e : s_->c1)
```
Comment (rec) [inline src/beast/beast/nudb/store.h]: braces around this multiline for statement please.  

Or you could just make the second statement into one line, it gains nothing from being on two lines.

File: src/beast/beast/nudb/detail/gentex.h [hunk 3]
```diff
@@ -73,7 +73,7 @@ gentex_t<_>::lock()
 
 template <class _>
 void
-gentex_t<_>::unlock()
+gentex_t<_>::finish()
 {
     std::unique_lock<
```
Comment (rec) [inline src/beast/beast/nudb/detail/gentex.h]: You really 
need two 
lines for 
the 40 
characters 
used to 
define this 
single variable?

File: src/beast/beast/nudb/win32_file.h (hunk 1 above)
Comment (vinniefalco) [inline src/beast/beast/nudb/win32_file.h]: @rec Yeah, its an unfortunate consequence of Microsoft's sloppy headers. If we define it for the whole project, then code which needs more than what's in the lean and mean windows headers will break.

File: src/beast/beast/nudb/store.h (hunk 2 above)
Comment (vinniefalco) [inline src/beast/beast/nudb/store.h]: The body is only one statement

File: src/beast/beast/nudb/win32_file.h (hunk 1 above)
Comment (nbougalis) [inline src/beast/beast/nudb/win32_file.h]: "sloppy headers" is putting it mildly...

File: src/beast/beast/nudb/detail/gentex.h (hunk 3 above)
Comment (vinniefalco) [inline src/beast/beast/nudb/detail/gentex.h]: By splitting the declaration of the local variable we bring the ratio between the longest and second longest lines down to 1.125 from 1.444. The code is formatted using subjective rules to create eye-pleasing shapes which optimize the properties of foveal vision. Here's a side by side comparison of foveal optimized versus traditional formatting, with visual features highlighted in red:
http://i.imgur.com/qhIBL81.png

**Comment by ximinez** [comment]
Correct me if I'm wrong, but the names of the `gentex` and `genlock` classes are a little deceiving. `genlock` doesn't really lock anything in the traditional sense, but is more tracking state counters in the `gentex`. The `gentex` doesn't block in the `start` function, but only in the `finish` function, where it waits until all `genlock`s taken in its "generation" have finished what they're doing and unlock. Everything else is basically completely asynchronous.

I also note that `gen_` monotonically increases, but given that it's 64-bit, it's a safe bet that it'll never get the chance to roll over, and if it ever does, it's pretty much guaranteed that those generations will have finished by then.

**Comment by JoelKatz** [comment]
@ximinez It locks everything associated with a generation while that generation is in use. It does this by delaying the destruction of the generation until no thread could possibly be using it. It is to some extent more like a smart pointer than a lock.

**Comment by ximinez** [comment]
@JoelKatz Ok, I had the general shape of it correct. Thanks for clarifying.

**Comment by rec** [comment]
:+1:

**Pull request closed without merging**